File: request-reply2/src/broker/cliente.py
import random
import string
import threading
import time

import msgpack
import zmq

BROKER_URL = "tcp://broker:5555"
SUB_URL = "tcp://pubsub-proxy:5558"
USERNAME = "bot_01"
MIN_CHANNELS = 5
MIN_SUBSCRIPTIONS = 3


class LamportClock:
    def __init__(self):
        self._value = 0
        self._lock = threading.Lock()

    def tick(self):
        with self._lock:
            self._value += 1
            return self._value

    def merge(self, received):
        with self._lock:
            self._value = max(self._value, int(received or 0))
            return self._value

    @property
    def value(self):
        with self._lock:
            return self._value


def build_request(message_type, payload, lamport_clock):
    return {
        "type": message_type,
        "timestamp": time.time(),
        "logical_clock": lamport_clock.tick(),
        "payload": payload,
    }


def send_request(socket, message, lamport_clock):
    socket.send(msgpack.packb(message, use_bin_type=True))
    raw_reply = socket.recv()
    reply = msgpack.unpackb(raw_reply, raw=False)
    lamport_clock.merge(reply.get("logical_clock", 0))
    return reply


def fazer_login(socket, username, lamport_clock):
    while True:
        req = build_request("login", {"username": username}, lamport_clock)
        reply = send_request(socket, req, lamport_clock)
        print(f"[LOGIN] Resposta: {reply}")

        if reply.get("status") == "ok":
            print(f"[LOGIN] Login bem-sucedido como {username!r}.")
            return True

        erro = reply.get("payload", {}).get("message", "erro desconhecido")
        print(f"[LOGIN] Falha: {erro}. Tentando novamente em 3s...")
        time.sleep(3)


def listar_canais(socket, lamport_clock):
    req = build_request("list_channels", {}, lamport_clock)
    reply = send_request(socket, req, lamport_clock)
    print(f"[LISTAR CANAIS] Resposta: {reply}")

    if reply.get("status") == "ok":
        canais = reply.get("payload", {}).get("channels", [])
        print(f"[LISTAR CANAIS] Canais disponíveis: {canais}")
        return canais

    erro = reply.get("payload", {}).get("message", "erro desconhecido")
    print(f"[LISTAR CANAIS] Erro: {erro}")
    return []


def criar_canal(socket, username, nome_canal, lamport_clock):
    payload = {"username": username, "channel": nome_canal}
    req = build_request("create_channel", payload, lamport_clock)
    reply = send_request(socket, req, lamport_clock)
    print(f"[CRIAR CANAL] Resposta: {reply}")

    if reply.get("status") == "ok":
        print(f"[CRIAR CANAL] Canal {nome_canal!r} criado com sucesso.")
        return True

    erro = reply.get("payload", {}).get("message", "erro desconhecido")
    print(f"[CRIAR CANAL] Não foi possível criar o canal {nome_canal!r}: {erro}")
    return False


def publicar(socket, username, canal, mensagem, lamport_clock):
    payload = {
        "username": username,
        "channel": canal,
        "message": mensagem,
    }
    req = build_request("publish_message", payload, lamport_clock)
    reply = send_request(socket, req, lamport_clock)
    status = reply.get("status")
    if status != "ok":
        erro = reply.get("payload", {}).get("message", "erro desconhecido")
        print(f"[PUB-REQ] Falha ao publicar em {canal!r}: {erro}")
    else:
        print(f"[PUB-REQ] Publicação confirmada em {canal!r}.")


def generate_channel_name(username):
    suffix = "".join(random.choices(string.ascii_lowercase + string.digits, k=5))
    return f"{username}_{suffix}"


def generate_message():
    suffix = "".join(random.choices(string.ascii_lowercase + string.digits, k=8))
    return f"msg_{suffix}"


def listen_subscriptions(sub_socket, stop_event, lamport_clock):
    poller = zmq.Poller()
    poller.register(sub_socket, zmq.POLLIN)

    while not stop_event.is_set():
        events = dict(poller.poll(timeout=200))
        if sub_socket not in events:
            continue

        topic, raw_payload = sub_socket.recv_multipart()
        received_timestamp = time.time()
        payload = msgpack.unpackb(raw_payload, raw=False)
        lamport_clock.merge(payload.get("logical_clock", 0))

        channel = payload.get("channel", topic.decode("utf-8", errors="replace"))
        message = payload.get("message", "")
        sent_timestamp = payload.get("sent_timestamp", "?")

        print(
            "[SUB] "
            f"canal={channel} | mensagem={message} | "
            f"ts_envio={sent_timestamp} | ts_recebimento={received_timestamp} | lc={lamport_clock.value}",
            flush=True,
        )


def main():
    context = zmq.Context()
    lamport_clock = LamportClock()

    req_socket = context.socket(zmq.REQ)
    req_socket.connect(BROKER_URL)
    print(f"[CLIENTE] Conectado ao broker em {BROKER_URL}")

    sub_socket = context.socket(zmq.SUB)
    sub_socket.connect(SUB_URL)
    print(f"[CLIENTE] Conectado ao proxy Pub/Sub em {SUB_URL}")

    stop_event = threading.Event()
    listener = threading.Thread(target=listen_subscriptions, args=(sub_socket, stop_event, lamport_clock), daemon=True)
    listener.start()

    subscribed_channels = set()

    try:
        fazer_login(req_socket, USERNAME, lamport_clock)

        canais = listar_canais(req_socket, lamport_clock)
        if len(canais) < MIN_CHANNELS:
            novo_canal = generate_channel_name(USERNAME)
            if criar_canal(req_socket, USERNAME, novo_canal, lamport_clock):
                canais = listar_canais(req_socket, lamport_clock)

        while True:
            canais = listar_canais(req_socket, lamport_clock)
            if not canais:
                print("[CLIENTE] Nenhum canal disponível. Tentando novamente em 2s...")
                time.sleep(2)
                continue

            while len(subscribed_channels) < MIN_SUBSCRIPTIONS:
                candidatos = [c for c in canais if c not in subscribed_channels]
                if not candidatos:
                    break
                canal = random.choice(candidatos)
                sub_socket.setsockopt_string(zmq.SUBSCRIBE, canal)
                subscribed_channels.add(canal)
                print(f"[SUB] Inscrito no canal {canal!r}")

            for _ in range(10):
                canal = random.choice(canais)
                mensagem = generate_message()
                publicar(req_socket, USERNAME, canal, mensagem, lamport_clock)
                time.sleep(1)

    except KeyboardInterrupt:
        print("[CLIENTE] Encerrando por interrupção do usuário.")
    finally:
        stop_event.set()
        listener.join(timeout=1)
        req_socket.close(0)
        sub_socket.close(0)
        context.term()
        print("[CLIENTE] Encerrado.")


if __name__ == "__main__":
    main()

File: request-reply2/src/broker/java/src/main/java/Cliente.java
import org.zeromq.SocketType;
import org.zeromq.ZContext;
import org.zeromq.ZMQ;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;

public class Cliente {

    private static final String BROKER_URL = "tcp://broker:5555";
    private static final String SUB_URL = "tcp://pubsub-proxy:5558";

    private static final String USERNAME =
        System.getenv().getOrDefault("USERNAME", "bot_java_01");

    private static final int MIN_CHANNELS = 5;
    private static final int MIN_SUBSCRIPTIONS = 3;
    private static final int PUBLISH_BATCH = 10;
    private static final int PUBLISH_INTERVAL_MS = 1000;

    private static long logicalClock = 0;

    public static void main(String[] args) throws Exception {
        try (ZContext ctx = new ZContext()) {
            ZMQ.Socket req = ctx.createSocket(SocketType.REQ);
            req.connect(BROKER_URL);
            System.out.println("[CLIENTE-JAVA] Conectado ao broker em " + BROKER_URL);

            ZMQ.Socket sub = ctx.createSocket(SocketType.SUB);
            sub.connect(SUB_URL);
            System.out.println("[CLIENTE-JAVA] Conectado ao proxy Pub/Sub em " + SUB_URL);

            final boolean[] running = new boolean[] {true};

            Thread listener = new Thread(() -> listenSubscriptions(sub, running));
            listener.setDaemon(true);
            listener.start();

            Random random = new Random();
            Set<String> subscribedChannels = new HashSet<>();

            try {
                fazerLogin(req, USERNAME);

                List<String> canais = ensureMinimumChannels(req, USERNAME, random);

                while (true) {
                    canais = listarCanais(req);

                    if (canais.isEmpty()) {
                        System.out.println("[CLIENTE-JAVA] Nenhum canal disponível. Tentando novamente em 2s...");
                        Thread.sleep(2000);
                        continue;
                    }

                    ensureMinimumSubscriptions(sub, canais, subscribedChannels, random);

                    for (int i = 0; i < PUBLISH_BATCH; i++) {
                        String canal = canais.get(random.nextInt(canais.size()));
                        String mensagem = generateMessage(random);
                        publicar(req, USERNAME, canal, mensagem);
                        Thread.sleep(PUBLISH_INTERVAL_MS);
                    }
                }

            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                System.out.println("[CLIENTE-JAVA] Encerrado por interrupção.");
            } finally {
                running[0] = false;
                listener.join(1000);
            }
        }
    }

    private static void listenSubscriptions(ZMQ.Socket sub, boolean[] running) {
        sub.setReceiveTimeOut(200);

        while (running[0] && !Thread.currentThread().isInterrupted()) {
            byte[] topic = sub.recv(0);
            if (topic == null) {
                continue;
            }

            byte[] payloadRaw = sub.recv(0);
            if (payloadRaw == null) {
                continue;
            }

            try {
                Map<String, Object> payload = MsgHelper.unpack(payloadRaw);
                mergeClock(payload.get("logical_clock"));

                String channel = payload.getOrDefault(
                    "channel",
                    new String(topic, StandardCharsets.UTF_8)
                ).toString();

                String message = payload.getOrDefault("message", "").toString();
                Object sentTimestamp = payload.getOrDefault("sent_timestamp", "?");
                double receivedTimestamp = now();

                System.out.println(
                    "[SUB-JAVA] canal=" + channel
                        + " | mensagem=" + message
                        + " | ts_envio=" + sentTimestamp
                        + " | ts_recebimento=" + receivedTimestamp
                        + " | lc=" + logicalClock
                );
            } catch (Exception e) {
                System.out.println("[SUB-JAVA] Erro ao processar mensagem: " + e.getMessage());
            }
        }
    }

    private static List<String> ensureMinimumChannels(ZMQ.Socket socket, String username, Random random) throws Exception {
        List<String> canais = listarCanais(socket);

        while (canais.size() < MIN_CHANNELS) {
            String novoCanal = generateChannelName(username, random);
            boolean criado = criarCanal(socket, username, novoCanal);

            if (!criado) {
                break;
            }

            canais = listarCanais(socket);
        }

        return canais;
    }

    private static void ensureMinimumSubscriptions(
        ZMQ.Socket sub,
        List<String> channels,
        Set<String> subscribed,
        Random random
    ) {
        while (subscribed.size() < MIN_SUBSCRIPTIONS) {
            List<String> candidates = new ArrayList<>();

            for (String c : channels) {
                if (!subscribed.contains(c)) {
                    candidates.add(c);
                }
            }

            if (candidates.isEmpty()) {
                return;
            }

            String selected = candidates.get(random.nextInt(candidates.size()));
            sub.subscribe(selected.getBytes(StandardCharsets.UTF_8));
            subscribed.add(selected);

            System.out.println("[SUB-JAVA] Inscrito no canal " + selected);
        }
    }

    private static void fazerLogin(ZMQ.Socket socket, String username) throws Exception {
        while (true) {
            Map<String, Object> payload = new LinkedHashMap<>();
            payload.put("username", username);

            Map<String, Object> req = buildRequest("login", payload);
            Map<String, Object> reply = sendRequest(socket, req);

            System.out.println("[LOGIN-JAVA] Resposta: " + reply);

            if ("ok".equals(reply.get("status"))) {
                System.out.println("[LOGIN-JAVA] Login bem-sucedido como " + username + ".");
                return;
            }

            String erro = extractError(reply);
            System.out.println("[LOGIN-JAVA] Falha: " + erro + ". Tentando novamente em 3s...");
            Thread.sleep(3000);
        }
    }

    @SuppressWarnings("unchecked")
    private static List<String> listarCanais(ZMQ.Socket socket) throws Exception {
        Map<String, Object> req = buildRequest("list_channels", new LinkedHashMap<>());
        Map<String, Object> reply = sendRequest(socket, req);

        System.out.println("[LISTAR CANAIS-JAVA] Resposta: " + reply);

        if ("ok".equals(reply.get("status"))) {
            Map<String, Object> payload =
                (Map<String, Object>) reply.getOrDefault("payload", new HashMap<>());

            List<Object> rawChannels =
                (List<Object>) payload.getOrDefault("channels", new ArrayList<>());

            List<String> canais = new ArrayList<>();
            for (Object channel : rawChannels) {
                canais.add(channel.toString());
            }

            System.out.println("[LISTAR CANAIS-JAVA] Canais disponíveis: " + canais);
            return canais;
        }

        System.out.println("[LISTAR CANAIS-JAVA] Erro: " + extractError(reply));
        return new ArrayList<>();
    }

    private static boolean criarCanal(ZMQ.Socket socket, String username, String canal) throws Exception {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("username", username);
        payload.put("channel", canal);

        Map<String, Object> req = buildRequest("create_channel", payload);
        Map<String, Object> reply = sendRequest(socket, req);

        System.out.println("[CRIAR CANAL-JAVA] Resposta: " + reply);

        if ("ok".equals(reply.get("status"))) {
            System.out.println("[CRIAR CANAL-JAVA] Canal " + canal + " criado com sucesso.");
            return true;
        }

        System.out.println("[CRIAR CANAL-JAVA] Falha ao criar canal " + canal + ": " + extractError(reply));
        return false;
    }

    private static boolean publicar(ZMQ.Socket socket, String username, String canal, String mensagem) throws Exception {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("username", username);
        payload.put("channel", canal);
        payload.put("message", mensagem);

        Map<String, Object> req = buildRequest("publish_message", payload);
        Map<String, Object> reply = sendRequest(socket, req);

        if ("ok".equals(reply.get("status"))) {
            System.out.println("[PUB-REQ-JAVA] Publicação confirmada em " + canal + ".");
            return true;
        }

        System.out.println("[PUB-REQ-JAVA] Falha ao publicar em " + canal + ": " + extractError(reply));
        return false;
    }

    private static String generateChannelName(String username, Random random) {
        return username + "_" + randomSuffix(random, 5);
    }

    private static String generateMessage(Random random) {
        return "msg_" + randomSuffix(random, 8);
    }

    private static String randomSuffix(Random random, int size) {
        final String alphabet = "abcdefghijklmnopqrstuvwxyz0123456789";
        StringBuilder sb = new StringBuilder(size);

        for (int i = 0; i < size; i++) {
            sb.append(alphabet.charAt(random.nextInt(alphabet.length())));
        }

        return sb.toString();
    }

    @SuppressWarnings("unchecked")
    private static String extractError(Map<String, Object> reply) {
        Map<String, Object> payload =
            (Map<String, Object>) reply.getOrDefault("payload", new HashMap<>());

        return payload.getOrDefault("message", "erro desconhecido").toString();
    }

    private static synchronized void mergeClock(Object received) {
        long receivedValue = (received == null) ? 0 : ((Number) received).longValue();
        logicalClock = Math.max(logicalClock, receivedValue);
    }

    private static synchronized long tickClock() {
        logicalClock += 1;
        return logicalClock;
    }

    private static double now() {
        return System.currentTimeMillis() / 1000.0;
    }

    private static Map<String, Object> buildRequest(String type, Map<String, Object> payload) {
        Map<String, Object> req = new LinkedHashMap<>();
        req.put("type", type);
        req.put("timestamp", now());
        req.put("logical_clock", tickClock());
        req.put("payload", payload);
        return req;
    }

    private static Map<String, Object> sendRequest(ZMQ.Socket socket, Map<String, Object> req) throws Exception {
        socket.send(MsgHelper.pack(req));
        byte[] raw = socket.recv();
        Map<String, Object> reply = MsgHelper.unpack(raw);
        mergeClock(reply.get("logical_clock"));
        return reply;
    }
}